rali/rali/source/caffe_lmdb_record_reader.cpp

```diff
@@ -0,0 +1,283 @@
+
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+*/
+
+#include <cassert>
+#include <commons.h>
+#include <boost/filesystem.hpp>
+#include <boost/algorithm/string.hpp>
+#include <iostream>
+#include <string>
+#include <vector>
+#include <sstream>
+#include <fstream>
+#include <stdint.h>
+#include "caffe_lmdb_record_reader.h"
+
+using namespace std;
+using caffe_protos::Datum;
+
+namespace filesys = boost::filesystem;
+
+CaffeLMDBRecordReader::CaffeLMDBRecordReader():
+_shuffle_time("shuffle_time", DBG_TIMING)
+{
+    _sub_dir = nullptr;
+    _curr_file_idx = 0;
+    _current_file_size = 0;
+    _loop = false;
+    _shuffle = false;
+    _file_id = 0;
+    _last_rec = false;
+}
+
+unsigned CaffeLMDBRecordReader::count()
+{
+    if (_loop)
+        return _file_names.size();
+
+    int ret = ((int)_file_names.size() - _read_counter);
+    return ((ret < 0) ? 0 : ret);
+}
+
+Reader::Status CaffeLMDBRecordReader::initialize(ReaderConfig desc)
+{
+    auto ret = Reader::Status::OK;
+    _file_id = 0;
+    _folder_path = desc.path();
+    _path = desc.path();
+    _shard_id = desc.get_shard_id();
+    _shard_count = desc.get_shard_count();
+    _batch_count = desc.get_batch_size();
+    _loop = desc.loop();
+    _shuffle = desc.shuffle();
+    ret = folder_reading();
+    //shuffle dataset if set
+    _shuffle_time.start();
+    if( ret==Reader::Status::OK && _shuffle)
+        std::random_shuffle(_file_names.begin(), _file_names.end());
+    _shuffle_time.end();
+
+    return ret;
+
+}
+
+void CaffeLMDBRecordReader::incremenet_read_ptr()
+{
+    _read_counter++;
+    _curr_file_idx = (_curr_file_idx + 1) % _file_names.size();
+}
+size_t CaffeLMDBRecordReader::open()
+{
+    auto file_path = _file_names[_curr_file_idx]; // Get next file name
+    _last_id = file_path;
+    _current_file_size = _file_size[_file_names[_curr_file_idx]];
+    return _current_file_size;
+}
+
+size_t CaffeLMDBRecordReader::read(unsigned char *buf, size_t read_size)
```

Comment:
Codacy found an issue: [Check buffer boundaries if used in a loop including recursive loops (CWE-120, CWE-20).  ](https://app.codacy.com/gh/GPUOpen-ProfessionalCompute-Libraries/MIVisionX/pullRequest?prid=5976200)